File: src/pages/login.page.ts
import { Locator, expect } from '@playwright/test';
import { BasePage } from './base.page.js';

/**
 * Page object for the login page.
 * Handles authentication flow for the application.
 *
 * Selector Strategy (in order of preference):
 * 1. getByLabel - Accessible, semantic, resilient to UI changes
 * 2. getByPlaceholder - User-visible, relatively stable
 * 3. CSS attribute selectors - Fallback when semantic selectors unavailable
 * 4. CSS type selectors - Last resort
 */
export class LoginPage extends BasePage {
  public readonly path = '/';

  /**
   * Get the username input field.
   * Uses multiple selector strategies for resilience.
   */
  private get usernameInput(): Locator {
    return this.page
      .getByLabel('Username')
      .or(this.page.getByPlaceholder(/username/i))
      .or(this.page.locator('input[name="username"]'))
      .or(this.page.locator('input[type="text"]').first());
  }

  /**
   * Get the password input field.
   * Uses multiple selector strategies for resilience.
   */
  private get passwordInput(): Locator {
    return this.page
      .getByLabel('Password')
      .or(this.page.getByPlaceholder(/password/i))
      .or(this.page.locator('input[name="password"]'))
      .or(this.page.locator('input[type="password"]'));
  }

  /**
   * Get the login/submit button.
   * Matches common button patterns for login forms.
   */
  private get loginButton(): Locator {
    return this.page
      .getByRole('button', { name: /sign in|log in|login|submit/i })
      .or(this.page.locator('button[type="submit"]'))
      .or(this.page.locator('button').filter({ hasText: /sign in|log in|login/i }));
  }

  /**
   * Get the error message element.
   * Used to verify login failures.
   */
  private get errorMessage(): Locator {
    return this.page
      .getByRole('alert')
      .or(this.page.locator('[class*="error"]'))
      .or(this.page.locator('[class*="alert"]'));
  }

  /**
   * Perform login with provided credentials.
   *
   * @param username - The username to enter
   * @param password - The password to enter
   */
  public async login(username: string, password: string): Promise<void> {
    await this.fillInput(this.usernameInput, username, 'username input');
    await this.fillInput(this.passwordInput, password, 'password input');
    await this.clickElement(this.loginButton, 'login button');

    // Wait for navigation after login
    await this.page.waitForLoadState('domcontentloaded');
  }

  /**
   * Verify login was successful by checking we're no longer on login page.
   * This is used after login to confirm authentication worked.
   */
  public async verifyLoginSuccessful(): Promise<void> {
    // Wait for URL to change from login page or for a dashboard element
    await expect(this.page).not.toHaveURL(/login/i, { timeout: 10000 });

    // Also verify login form is no longer visible
    await expect(this.loginButton).toBeHidden({ timeout: 5000 });
  }

  /**
   * Verify login error message is displayed.
   *
   * @param expectedMessage - The expected error message text
   */
  public async verifyLoginError(expectedMessage: string): Promise<void> {
    await expect(this.errorMessage).toContainText(expectedMessage);
  }
}

File: python-behave/pages/login_page.py
"""
Login page object.

Handles authentication flows for the demo application.
"""
from playwright.sync_api import Page
from .base_page import BasePage
from core.logger import get_logger

logger = get_logger(__name__)


class LoginPage(BasePage):
    """Page object for the login page."""

    # Selectors
    USERNAME_INPUT = 'input[id="username"]'
    PASSWORD_INPUT = 'input[id="password"]'
    LOGIN_BUTTON = 'button[type="submit"]'
    ERROR_MESSAGE = '[class*="error"], [class*="alert"], [role="alert"]'

    @property
    def url_path(self) -> str:
        return "/"

    def login(self, username: str, password: str) -> "LoginPage":
        """
        Perform login with given credentials.

        Args:
            username: Login username
            password: Login password

        Returns:
            Self for method chaining
        """
        logger.info(f"Logging in as {username}")

        self.fill(self.USERNAME_INPUT, username)
        self.fill(self.PASSWORD_INPUT, password)
        self.click(self.LOGIN_BUTTON)

        return self

    def login_as_default_user(self) -> "LoginPage":
        """Login with default credentials from config."""
        return self.login(self.config.username, self.config.password)

    def is_logged_in(self) -> bool:
        """Check if user is logged in (sidebar visible)."""
        # After login, sidebar with projects should be visible
        return self.is_visible('[class*="sidebar"]', timeout=5000) or \
               self.is_visible('nav', timeout=5000)

    def has_error(self) -> bool:
        """Check if login error is displayed."""
        return self.is_visible(self.ERROR_MESSAGE, timeout=2000)

    def get_error_message(self) -> str:
        """Get error message text."""
        if self.has_error():
            return self.get_text(self.ERROR_MESSAGE)
        return ""

    def is_on_login_page(self) -> bool:
        """Check if still on login page."""
        return self.is_visible(self.LOGIN_BUTTON, timeout=2000)
